packages/shared/friends/sagas.ts

```diff
@@ -799,3 +822,246 @@ function logAndTrackError(message: string, e: any) {
     stack: '' + e
   })
 }
+
+// Join or create channel
+async function* handleJoinOrCreateChannel(action: JoinOrCreateChannel) {
+  const client: SocialAPI | null = getSocialClient(store.getState())
+  if (!client) return
+
+  const channelId = action.payload.channelId
+  const ownId = client.getUserId()
+
+  // get or create channel
+  const { created, conversation } = await client.getOrCreateChannel(channelId, [ownId])
+
+  const channel: ChannelInfoPayload = {
+    name: conversation.name!,
+    channelId: conversation.id,
+    unseenMessages: 0,
+    lastMessageTimestamp: undefined,
+    memberCount: 1,
+    description: '',
+    joined: true,
+    muted: false
+  }
+
+  if (!created) {
+    await client.joinChannel(conversation.id)
+
+    const joinedChannel = client.getChannel(conversation.id)
+
+    channel.unseenMessages = joinedChannel?.unreadMessages?.length || 0
+    channel.lastMessageTimestamp = joinedChannel?.lastEventTimestamp
+    channel.memberCount = joinedChannel?.userIds?.length || 0
+  }
+
+  // parse channel info
+  const channelsInfo: ChannelsInfoPayload = {
+    channelsInfoPayload: [channel]
+  }
+
+  // send confirmation message to unity
+  getUnityInstance().JoinChannelConfirmation(channelsInfo)
+}
+
+// Create channel
+export async function createChannel(request: CreateChannelPayload) {
+  const client: SocialAPI | null = getSocialClient(store.getState())
+  if (!client) return
+
+  const channelId = request.channelId
+  const ownId = client.getUserId()
+
+  if (CHANNEL_RESERVED_IDS.includes(channelId)) {
+    joinChannelError(channelId, 'Reserved name')
+    return
+  }
+
+  if (!validateRegexChannelId(channelId)) {
+    joinChannelError(channelId, 'Bad regex')
+    return
+  }
```

Comment:
Can we extract these validations and use from the previous join or create command as well? Also to unify error messages.

Reply:
Yes, this should be done from the client side